BUG: Add cleanup to SegmentEditorEffect

tom-osika opened the pull request:
Closes #7392 

Adds a `cleanup` method to the various editor effects classes that disconnects any signals / slots. This is necessary before calling `delete`, as the connection may keep the object alive otherwise

Here's the output of the patch / code linked in the original issue after applying this fix:

```
Deleting AbstractScriptedSegmentEditorAutoCompleteEffect: <class 'SegmentEditorEffects.SegmentEditorFillBetweenSlicesEffect.SegmentEditorFillBetweenSlicesEffect'>
Deleting AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorFillBetweenSlicesEffect.SegmentEditorFillBetweenSlicesEffect'>
Deleting AbstractScriptedSegmentEditorAutoCompleteEffect: <class 'SegmentEditorEffects.SegmentEditorGrowFromSeedsEffect.SegmentEditorGrowFromSeedsEffect'>
Deleting AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorGrowFromSeedsEffect.SegmentEditorGrowFromSeedsEffect'>
Deleting AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorHollowEffect.SegmentEditorHollowEffect'>
Deleting AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorIslandsEffect.SegmentEditorIslandsEffect'>
Deleting AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorLogicalEffect.SegmentEditorLogicalEffect'>
Deleting AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorMarginEffect.SegmentEditorMarginEffect'>
Deleting AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorMaskVolumeEffect.SegmentEditorMaskVolumeEffect'>
Switch to module:  ""
Deleting AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorDrawEffect.SegmentEditorDrawEffect'>
Deleting AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorLevelTracingEffect.SegmentEditorLevelTracingEffect'>
Deleting AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorSmoothingEffect.SegmentEditorSmoothingEffect'>

```

Unlike before, the `FillBetweenSlices` and `GrowFromSeeds` effects are deleted. All of the non-abstract effects are deleted (The C++ ones aren't printed out because the patch only applied to the python)

jcfr commented on Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorEffects/AbstractScriptedSegmentEditorAutoCompleteEffect.py:
```diff
@@ -66,6 +66,10 @@ def __del__(self):
+    def cleanup(self):
+        self.delayedAutoUpdateTimer.disconnect("timeout()", self.onPreview)
+
+
```
```suggestion
```

tom-osika replied:
Updated the last commit to include this change

lassoan commented on Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorEffects/AbstractScriptedSegmentEditorAutoCompleteEffect.py:
```diff
@@ -66,6 +66,9 @@ def __del__(self):
         self.delayedAutoUpdateTimer.stop()
         self.observeSegmentation(False)
 
+    def cleanup(self):
```
add a comment that describes why this is needed (if there is no comment and no failing test then the next developer may remove this call when he does some refactoring)

lassoan commented on Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorEffects/AbstractScriptedSegmentEditorEffect.py:
```diff
@@ -53,6 +53,9 @@ def register(self):
         effectFactorySingleton.registerEffect(self.scriptedEffect)
 
+    def cleanup(self):
+        pass
```
describe in a comment when this is called and what it can be used for

lassoan commented on Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorEffects/SegmentEditorThresholdEffect.py:
```diff
@@ -67,6 +67,9 @@ def __init__(self, scriptedEffect):
         self.enableViewInteractions = True
 
+    def cleanup(self):
+        self.timer.disconnect("timeout()", self.preview)
```
add a short comment describing why this line was needed

lassoan commented on Modules/Loadable/Segmentations/EditorEffects/qSlicerSegmentEditorAbstractEffect.h:
```diff
@@ -193,6 +193,8 @@ class Q_SLICER_SEGMENTATIONS_EFFECTS_EXPORT qSlicerSegmentEditorAbstractEffect :
   /// Default behavior is to deactivate the effect if not in view mode.
   virtual void interactionNodeModified(vtkMRMLInteractionNode* interactionNode);
 
+  Q_INVOKABLE virtual void cleanup() { };
```
Add method documentation

lassoan commented on Modules/Loadable/Segmentations/EditorEffects/qSlicerSegmentEditorScriptedEffect.h:
```diff
@@ -119,6 +119,8 @@ class Q_SLICER_SEGMENTATIONS_EFFECTS_EXPORT qSlicerSegmentEditorScriptedEffect
   void layoutChanged() override;
   /// Let the effect know that the interaction node is modified
   void interactionNodeModified(vtkMRMLInteractionNode* interactionNode) override;
+  /// Remove any active signal / slot connections before calling delete
```
```suggestion
  /// Release any resources, event observers, Qt signal / slot connections before calling delete
```

lassoan reviewed (commented):
Thanks you, this looks good, I just added a few comments on how to improve documentation/readability of the code.

tom-osika commented:
> Thanks you, this looks good, I just added a few comments on how to improve documentation/readability of the code.

Thanks Andras. See the most recent commit (3563ffc2df6841439f5dacc5b409380434951ee0). That should have all of the documentation changes

jcfr commented on Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorEffects/AbstractScriptedSegmentEditorEffect.py:
```diff
@@ -53,6 +53,22 @@ def register(self):
 
+    def cleanup(self):
+        """
+        Release any resources, event observers, Qt signal / slot connections
+        before object is garbage collected.
+
+
+        Subclasses should override this method to disconnect any signals and
+        slots specific to that subclass. This should be called right before the
+        object is garbage collected (or before calling delete
+        on the c++ side). This is necessary because any signal / slot connections that remain
+        will prevent the object from being garbage collected / deleted.
+        Release those resources here.
+        See https://github.com/Slicer/Slicer/issues/7392 for more info
```
```suggestion
        Clean up resources, event observers, and Qt signal/slot connections 
        to ensure proper object deletion.

        Subclasses should override this method to disconnect any subclass-specific 
        signals and slots. This method should be called before the object is 
        garbage collected or explicitly deleted on the C++ side. 
        
        Failing to disconnect signals/slots may prevent the object from being 
        garbage collected, leading to memory leaks. 

        For more details, see: https://github.com/Slicer/Slicer/issues/7392
```

jcfr commented on Modules/Loadable/Segmentations/EditorEffects/qSlicerSegmentEditorAbstractEffect.h:
```diff
@@ -192,6 +192,10 @@ class Q_SLICER_SEGMENTATIONS_EFFECTS_EXPORT qSlicerSegmentEditorAbstractEffect :
   /// Let the effect know that the interaction node is modified.
   /// Default behavior is to deactivate the effect if not in view mode.
   virtual void interactionNodeModified(vtkMRMLInteractionNode* interactionNode);
+  /// Release any resources, event observers, Qt signal / slot connections before calling delete.
+  /// This is necessary as any active signal / slot connections will prevent destruction of the object
+  /// See https://github.com/Slicer/Slicer/issues/7392 for more info
```
```suggestion
  /// Clean up resources, event observers, and Qt signal/slot connections before deletion.
  ///
  /// This ensures proper object destruction, as active signal/slot connections 
  /// can prevent the object from being deleted. Subclasses should override this 
  /// method to handle additional cleanup as needed.
  ///
  /// For more details, see: https://github.com/Slicer/Slicer/issues/7392
```

jcfr commented on Modules/Loadable/Segmentations/EditorEffects/qSlicerSegmentEditorScriptedEffect.h:
```diff
@@ -119,6 +119,8 @@ class Q_SLICER_SEGMENTATIONS_EFFECTS_EXPORT qSlicerSegmentEditorScriptedEffect
   void layoutChanged() override;
   /// Let the effect know that the interaction node is modified
   void interactionNodeModified(vtkMRMLInteractionNode* interactionNode) override;
+  /// Release any resources, event observers, Qt signal / slot connections before calling delete
```
```suggestion
  /// Clean up resources, event observers, and Qt signal/slot connections before deletion.
```

I also though about using 

```
  /// \copydoc qSlicerSegmentEditorScriptedEffect::cleanup()
```

.. but this is not done anywhere else.
See https://www.doxygen.nl/manual/commands.html#cmdcopydoc

jcfr commented on Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorEffects/SegmentEditorThresholdEffect.py:
```diff
@@ -67,6 +67,13 @@ def __init__(self, scriptedEffect):
         self.enableViewInteractions = True
 
+    def cleanup(self):
+        # Just disconnect the timer signal / slot so we can
+        # garbage collect this object
+        # See cleanup() docstring in parent class or the following issue:
+        # https://github.com/Slicer/Slicer/issues/7392
```
```suggestion
        # Disconnect the timer signal to allow proper garbage collection.
        #
        # This prevents lingering signal/slot connections from keeping 
        # the object alive. For more details, see the parent class 
        # cleanup() docstring or the following issue:
        # https://github.com/Slicer/Slicer/issues/7392
```

jcfr commented on Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorEffects/AbstractScriptedSegmentEditorAutoCompleteEffect.py:
```diff
@@ -66,6 +66,13 @@ def __del__(self):
         self.observeSegmentation(False)
 
+    def cleanup(self):
+        # Just disconnect the timer signal / slot so we can
+        # garbage collect this object
+        # See cleanup() docstring in parent class or the following issue:
+        # https://github.com/Slicer/Slicer/issues/7392
```
```suggestion
        # Disconnect the timer signal to allow proper garbage collection.
        #
        # This prevents lingering signal/slot connections from keeping 
        # the object alive. For more details, see the parent class 
        # cleanup() docstring or the following issue:
        # https://github.com/Slicer/Slicer/issues/7392
```

jcfr reviewed (commented):
Thanks for the polishing your contribution and addressing the review comments :pray: :100: 

After integrating the very minor suggestions, it will be good for :rocket:

tom-osika commented:
> Thanks for the polishing your contribution and addressing the review comments 🙏 💯
> 
> After integrating the very minor suggestions, it will be good for 🚀

Thanks for the suggestions. I've applied them and squashed into the most recent commit

jcfr reviewed (approved):
I was able to test locally and I will move forward with `Squash & Merge` specifying the following commit message:

```
BUG: Ensure proper cleanup of Segment Editor effects

- Introduce `cleanup()` method in `qSlicerSegmentEditorAbstractEffect` and related classes
- Update `qMRMLSegmentEditorWidgetPrivate` destructor to call `cleanup()` before effect deletion
- Implemente `cleanup()` in Python scripted effects to disconnect signals and slots, preventing memory leaks

This improves memory management and prevents objects from being kept alive unnecessarily.

Co-authored-by: Andras Lasso <lasso@queensu.ca>
Co-authored-by: Jean-Christophe Fillion-Robin <jchris.fillionr@kitware.com>
```

jcfr reviewed (changes requested):
> [!WARNING]
> Comparing the list of effect instantiated vs deleted, the `SegmentEditorThresholdEffect` does not seem to be deleted

Here is the output after instrumenting both the `__init__` and `__del__`:

```
Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorDrawEffect.SegmentEditorDrawEffect'>
Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorFillBetweenSlicesEffect.SegmentEditorFillBetweenSlicesEffect'>
Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorGrowFromSeedsEffect.SegmentEditorGrowFromSeedsEffect'>
Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorHollowEffect.SegmentEditorHollowEffect'>
Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorIslandsEffect.SegmentEditorIslandsEffect'>
Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorLevelTracingEffect.SegmentEditorLevelTracingEffect'>
Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorLogicalEffect.SegmentEditorLogicalEffect'>
Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorMarginEffect.SegmentEditorMarginEffect'>
Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorMaskVolumeEffect.SegmentEditorMaskVolumeEffect'>
Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorSmoothingEffect.SegmentEditorSmoothingEffect'>
Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorThresholdEffect.SegmentEditorThresholdEffect'>

Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorDrawEffect.SegmentEditorDrawEffect'>
Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorFillBetweenSlicesEffect.SegmentEditorFillBetweenSlicesEffect'>
Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorGrowFromSeedsEffect.SegmentEditorGrowFromSeedsEffect'>
Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorHollowEffect.SegmentEditorHollowEffect'>
Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorIslandsEffect.SegmentEditorIslandsEffect'>
Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorLevelTracingEffect.SegmentEditorLevelTracingEffect'>
Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorLogicalEffect.SegmentEditorLogicalEffect'>
Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorMarginEffect.SegmentEditorMarginEffect'>
Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorMaskVolumeEffect.SegmentEditorMaskVolumeEffect'>
Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorSmoothingEffect.SegmentEditorSmoothingEffect'>
Instantiating AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorThresholdEffect.SegmentEditorThresholdEffect'>

availableEffectNames: ('Paint', 'Erase', 'Scissors', 'Draw', 'Fill between slices', 'Grow from seeds', 'Hollow', 'Islands', 'Level tracing', 'Logical operators', 'Margin', 'Mask volume', 'Smoothing', 'Threshold')
# of availableEffectNames: 14
effectCount: 15
Switch to module:  ""
Deleting AbstractScriptedSegmentEditorAutoCompleteEffect: <class 'SegmentEditorEffects.SegmentEditorFillBetweenSlicesEffect.SegmentEditorFillBetweenSlicesEffect'>
Deleting AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorFillBetweenSlicesEffect.SegmentEditorFillBetweenSlicesEffect'>
Deleting AbstractScriptedSegmentEditorAutoCompleteEffect: <class 'SegmentEditorEffects.SegmentEditorGrowFromSeedsEffect.SegmentEditorGrowFromSeedsEffect'>
Deleting AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorGrowFromSeedsEffect.SegmentEditorGrowFromSeedsEffect'>
Deleting AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorHollowEffect.SegmentEditorHollowEffect'>
Deleting AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorIslandsEffect.SegmentEditorIslandsEffect'>
Deleting AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorLogicalEffect.SegmentEditorLogicalEffect'>
Deleting AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorMarginEffect.SegmentEditorMarginEffect'>
Deleting AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorEffects.SegmentEditorMaskVolumeEffect.SegmentEditorMaskVolumeEffect'>
Switch to module:  ""
Deleting AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorDrawEffect.SegmentEditorDrawEffect'>
Deleting AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorLevelTracingEffect.SegmentEditorLevelTracingEffect'>
Deleting AbstractScriptedSegmentEditorEffect: <class 'SegmentEditorSmoothingEffect.SegmentEditorSmoothingEffect'>
```

tom-osika commented:
> Warning
> 
> Comparing the list of effect instantiated vs deleted, the `SegmentEditorThresholdEffect` does not seem to be deleted
> ...

Good catch. Thanks for taking care of that JC
